File: Arrays & Strings/#31 - Next Permutation - Medium/nextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int ind = -1;
        // Step 1: Find the break point
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                ind = i;
                break;
            }
        }
        // If break point does not exist, reverse the whole array
        if (ind == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }
        // Step 2: Find element just greater than nums[ind] and swap
        for (int i = n - 1; i > ind; i--) {
            if (nums[i] > nums[ind]) {
                swap(nums[i], nums[ind]);
                break;
            }
        }
        // Step 3: Reverse the right half
        reverse(nums.begin() + ind + 1, nums.end());
    }
};

File: Arrays & Strings/#31 - Next Permutation - Medium/nextPermutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        ind = -1
        # Step 1: Find the break point
        for i in range(n-2, -1, -1):
            if nums[i] < nums[i + 1]:
                ind = i
                break
        # If break point does not exist, reverse the whole array
        if ind == -1:
            nums.reverse()
            return
        # Step 2: Find the next greater element and swap
        for i in range(n - 1, ind, -1):
            if nums[i] > nums[ind]:
                nums[i], nums[ind] = nums[ind], nums[i]
                break
        # Step 3: Reverse the right half
        nums[ind+1:] = reversed(nums[ind+1:])
